docs/src/prototypes/mentions/MentionsEditor.tsx

```diff
@@ -0,0 +1,160 @@
+  args?: string
+}
+
+const editorStyle: React.CSSProperties = {
+  backgroundColor: '#eee',
+  borderRadius: '5px',
+  border: '1px dashed grey',
+  padding: '5px',
+  minHeight: '100px',
+  outline: 0,
+}
+
+const stateReducer = (state: MentionsEditorState, action: MentionsEditorAction) => {
+  switch (action.type) {
+    case 'OPEN':
+      return { ...state, open: true }
+    case 'RESET_UPDATE_FLAG':
+      return { ...state, shouldUpdate: false }
+    case 'UPDATE_AND_CLOSE':
+      return {
+        ...state,
+        shouldUpdate: true,
+        open: false,
+        searchQuery: '',
+        selectedItem: action.args,
+      }
+    case 'SET_SEARCH_QUERY':
+      return { ...state, searchQuery: action.args }
+    default:
+      throw new Error(`Action ${action.type} is not supported`)
+  }
+}
+
+const MentionsEditor: React.FunctionComponent<
+  MentionsContainerProps & { render: (props: MentionsContainerProps) => React.ReactNode }
+> = props => {
+  const { render, ...rest } = props
+  const contendEditableRef = React.useRef(null)
+  const [state, dispatch] = React.useReducer(stateReducer, {
+    shouldUpdate: false,
+    open: false,
+    searchQuery: '',
+    selectedItem: '',
+  })
+
+  React.useEffect(
+    () => {
+      if (!state.shouldUpdate) {
+        return
+      }
+
+      tryFocusEditor()
+
+      // after the wrapped component is closed the value of the search query is inserted in the editor at cursor position
+      insertTextAtCursorPosition(state.selectedItem)
+      dispatch({ type: 'RESET_UPDATE_FLAG' })
+    },
+    [state.shouldUpdate],
+  )
+
+  const handleEditorKeyUp = (e: React.KeyboardEvent) => {
+    if (!state.open && e.shiftKey && keyboardKey.getCode(e) === keyboardKey.AtSign) {
+      dispatch({ type: 'OPEN' })
+    }
+  }
+
+  const handleOpenChange = (e: React.SyntheticEvent, { open }: MentionsContainerProps) => {
+    if (!open) {
+      resetStateAndUpdateEditor()
+    }
+  }
+
+  const handleSelectedChange = (
+    e: React.SyntheticEvent,
+    { searchQuery }: MentionsContainerProps,
+  ) => {
+    resetStateAndUpdateEditor(searchQuery)
+  }
+
+  const handleSearchQueryChange = (
+    e: React.SyntheticEvent,
+    { searchQuery }: MentionsContainerProps,
+  ) => {
+    dispatch({ type: 'SET_SEARCH_QUERY', args: searchQuery })
+  }
+
+  const handleInputKeyDown = (e: React.KeyboardEvent) => {
+    const keyCode = keyboardKey.getCode(e)
+    switch (keyCode) {
+      case keyboardKey.Backspace: // 8
+        if (state.searchQuery === '') {
+          resetStateAndUpdateEditor()
+        }
+        break
+      case keyboardKey.Escape: // 27
+        resetStateAndUpdateEditor()
+        break
+    }
+  }
+
+  const resetStateAndUpdateEditor = (searchQueryArg?: string) => {
+    if (state.open) {
+      dispatch({ type: 'UPDATE_AND_CLOSE', args: searchQueryArg || state.searchQuery })
+    }
+  }
+
+  const tryFocusEditor = () => _.invoke(contendEditableRef.current, 'focus')
```

Comment:
This is used single time, we can omit it...

Reply:
done, thx